GRDBSnapshotTesting/Sources/Dump/DatabaseDumper.swift

```diff
@@ -64,34 +71,53 @@ struct DatabaseDumper {
         
         guard !views.isEmpty else { return }
         
-        lines.append(contentsOf: [header("VIEWS")])
+        lines.append(contentsOf: [header("VIEWS"), emptyLine])
         
         let createViews = views
             .map { $0.sql }
-            .flatMap { [$0] + [""] }
+            .flatMap { [$0] + [emptyLine] }
         
         lines.append(contentsOf: createViews)
     }
     
     private func dumpData(lines: inout [String]) throws {
         let tables = try database.read(SQLiteMaster.Requests.tables.fetchAll)
         
-        lines.append(contentsOf: [header("DATA")])
+        lines.append(contentsOf: [header("DATA"), emptyLine])
         
-        try tables.forEach { table in
-            lines.append("")
+        guard !tables.isEmpty else {
+            lines.append(contentsOf: [placeholder("NO DATA"), emptyLine])
             
+            return
+        }
+        
+        try tables.forEach { table in
             try dumpTable(table.name, lines: &lines)
+            
+            lines.append(emptyLine)
         }
+        
+        // Remove last added empty line
+        lines.removeLast()
     }
     
     private func dumpTable(_ table: String, lines: inout [String]) throws {
-        let rows = try database.read { try Row.fetchAll($0, sql: "SELECT * FROM \(table)") }
+        let rows = try database.read { db -> [Row] in
+            let primaryKey = try db.primaryKey(table)
+            let order = primaryKey.columns
+                .map { $0.quotedDatabaseIdentifier }
```

Comment:
I regret having suggested using `quotedDatabaseIdentifier`, because GRDB documents that the SQL it generates should no be relied upon. This means that `quotedDatabaseIdentifier`may change implementation without notice, and this could break snapshots.

So maybe quoting should instead be performed by GRDBSnapshotTesting itself.

Reply:
👍 I'll reimplement it